src/ElasticSearch/Parameters/Builder/SearchBuilder.php

```diff
@@ -227,52 +230,57 @@ protected function getSearchQueryOptions() {
   }
 
   /**
-   * Return a full text search query.
+   * Turn the given search keys into a lucene object structure.
    *
-   * TODO: better handling of parse modes.
-   *
-   * @param string[] $keys
-   *   Array of search keys, each one a string.
+   * @param array $keys
+   *   Search keys, in the format described by
+   *   \Drupal\search_api\ParseMode\ParseModeInterface::parseInput().
    * @param \Drupal\search_api\ParseMode\ParseModeInterface $parse_mode
-   *   Parse mode in use.
+   *   Search API parse mode.
+   * @param bool $fuzzy
+   *   Enable fuzzy support or not.
    *
-   * @return string
-   *   Full text search query string.
+   * @return \MakinaCorpus\Lucene\AbstractQuery
+   *   Return a lucene query object.
    */
-  protected function flattenKeys(array $keys, ParseModeInterface $parse_mode = NULL) {
+  protected function flattenKeys(array $keys, ParseModeInterface $parse_mode = NULL, $fuzzy = TRUE) {
+    // Grab the conjunction and negation properties if present.
     $conjunction = isset($keys['#conjunction']) ? $keys['#conjunction'] : 'AND';
     $negation = !empty($keys['#negation']);
-    $values = [];
 
-    foreach ($keys as $key_nr => $key) {
-      // We cannot use \Drupal\Core\Render\Element::children() anymore because
-      // $keys is not a valid render array.
-      if ($key_nr[0] === '#' || !$key) {
-        continue;
-      }
+    // Create a top level query.
+    $query = (new TermCollectionQuery())
+      ->setOperator($conjunction);
+    if ($negation) {
+      $query->setExclusion(Query::OP_PROHIBIT);
+    }
+
+    // Filter out top level properties beginning with '#'.
```

Comment:
Why do we do this?

Reply:
Because of how we get `$keys` from Search API module. As per docblock, `$keys` is documented in `\Drupal\search_api\ParseMode\ParseModeInterface::parseInput()`. Keys can be nested, which each level optionally defining a `#conjunction` and/or `#negation`, we want to loop over the keys (string or array) themselves only, after saving the property values above, we don't care about them. It's kind of a gnarly data structure I know. D7 arrays live on!

Reply:
Ah, got it! Since $keys is defined in ParseModeInterface, should we file an issue to add a "filter" method to that interface and implementation? Either way, certainly doesn't block this PR, it just seems a little confusing to jump around like that.